cadence-macros/tests/lib.rs

```diff
@@ -2,151 +2,125 @@ use cadence::{SpyMetricSink, StatsdClient};
 use cadence_macros::{
     statsd_count, statsd_distribution, statsd_gauge, statsd_histogram, statsd_meter, statsd_set, statsd_time,
 };
-use std::io;
-use std::sync::{Arc, Mutex, Once};
-
-/// Underlying writer to be used by a `SpyMetricSink`
-static mut WRITER: Option<Arc<Mutex<TestWriter>>> = None;
-
-/// Control initialization of the underlying writer
-static WRITER_INIT: Once = Once::new();
-
-/// `Write` implementation to use with a SpyMetricSink for testing
-///
-/// This write implementation converts all incoming writes to strings
-/// (assuming utf-8 encoding) and stores them in a vector that grows
-/// without bound.
-struct TestWriter {
-    saved: Vec<String>,
-}
-
-impl TestWriter {
-    fn new() -> Self {
-        TestWriter { saved: Vec::new() }
-    }
-
-    /// Get the underlying vector used for storage
-    fn storage(&self) -> &Vec<String> {
-        &self.saved
-    }
-}
+use crossbeam_channel::Receiver;
+use std::collections::HashSet;
+use std::sync::{Arc, Once};
 
-impl io::Write for TestWriter {
-    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        // test code, this should never be invalid utf-8
-        let s = String::from_utf8(buf.to_vec()).unwrap();
-        self.saved.push(s);
-        Ok(buf.len())
-    }
+/// Channel to be used to inspect metrics written to a `SpyMetricSink`
+static mut RX: Option<Arc<Receiver<Vec<u8>>>> = None;
```

Comment:
`static mut` has some super powers that very easily shoot unsuspecting feet. Try `static UnsafeCell<T>` instead. You may also be able to use another `std` type that provides interior mutability and a safe API, like `Cell` or `RefCell`, depending on your needs.

I know this is just in the tests and it isn't fundamentally different than the previous code. Still worth pointing out, IMHO.

Reply:
Good points. I'm going to spend some time seeing if I can solve this in a reasonable way with `UnsafeCell`. If I can't come up with anything, I'll open an issue to address this in the future since the global client used by macros uses the same approach.

Reply:
I'm going to merge this as-is and will fix this and the global state used in macros in a follow up PR.